add ec3 clients

RytisSapka opened the pull request:
Added EC3 Clients, unit and integration tests.

MindaugasButkus commented on src/reporting/clients/IEC3ConfigurationsClient.ts:
```diff
@@ -0,0 +1,88 @@
+
+export interface IEC3ConfigurationsClient {
+  /**
+   * Gets all EC3 Configurations for an iModel. This method returns the full list of EC3 Configurations.
```
EC3 configurations have nothing to do with iModels. They only refer to Reports. See https://developer.bentley.com/apis/carbon-calculation/operations/create-ec3-configuration/

RytisSapka replied:
removed the ```for an iModel``` parts

MindaugasButkus replied:
Is this comment resolved?

![image](https://user-images.githubusercontent.com/87982008/202456368-58df42dd-37b5-479e-ac4d-aefc9103bf6c.png)

RytisSapka replied:
changed all iModel references to project references.

MindaugasButkus commented on src/test/MappingsClient.test.ts:
```diff
@@ -214,7 +214,7 @@ describe("mappings Client", () => {
     mapping = await mappingsClientNewBase.updateMapping("auth", "iModelId", "mappingId", newMapping);
     expect(fetchStub.calledWith(
-      "https://api.bentley.com/insights/reporting/datasources/imodels/iModelId/mappings/mappingId",
+      "BASE/datasources/imodels/iModelId/mappings/mappingId",
```
What happened here?

MindaugasButkus replied:
You are reusing the same `fetchStub` multiple times in this test, so you do not know which method called `fetchStub` with what parameters.

RytisSapka replied:
I am reusing the fetch stub. I first fetch with the client that has the default base path ant then with the client that has ```BASE``` base path. When writing tests I noticed that this expect was not using the ```BASE``` path so I changed it (was epecting the same thing twice). Overall these tests should be remade using ```getcall(0).args``` because right now if they fail the message is ```Expected false be eqial to true```. Not sure if this should be done in this PR though.

MindaugasButkus replied:
What I wanted to point out that this change does not make sense. The test was not failing with the wrong expectation.

RytisSapka replied:
Now changing base path is being handled in another test

MindaugasButkus commented on src/test/EC3JobsClient.test.ts:
```diff
@@ -0,0 +1,76 @@
+import { CarbonUploadState, EC3JobCreate } from "../reporting/interfaces/EC3Jobs";
+use(chaiAsPromised);
+
+describe("EC3 Jobs Client Client", () => {
```
Duplicate `Client`. Also, make it match the class name (without spaces).

MindaugasButkus commented on src/integration-test/integration/ReportsClient.test.ts:
```diff
@@ -105,14 +105,6 @@ describe("Reports Client", () => {
     expect(report.description).to.be.eq("Updated");
   });
 
-  it("Reports - Get all non deleted", async () => {
```
Why did you remove this test?

RytisSapka replied:
Not sure how to correctly test this because if other tests create reports this test becomes polluted with them. Perhaps I should just check if the count is greater than 3?

RytisSapka replied:
Added test back.

MindaugasButkus commented on src/test/EC3ConfigurationsClient.test.ts:
```diff
@@ -0,0 +1,261 @@
+
+  beforeEach(() => {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    fetchStub = sinon.stub(EC3ConfigurationsClient.prototype, "fetchJSON" as any);
```
Have you tried stupping the `fetchJSON` method directly on the instance, not on the prototype?

RytisSapka replied:
stubbed on instance

MindaugasButkus commented on src/reporting/clients/IEC3ConfigurationsClient.ts:
```diff
@@ -0,0 +1,88 @@
+
+  /**
+   * Gets an async paged iterator for EC3 Configurations within the context of a Project.
+   * This method returns an iterator which loads pages of mappings as it is being iterated over.
```
The pages are most likely not `of mappings`.

RytisSapka replied:
Triple checked documentation

Guillar1 commented on src/reporting/clients/EC3ConfigurationsClient.ts:
```diff
@@ -0,0 +1,98 @@
+    super(basePath ?? CARBON_CALCULATION_BASE_PATH);
+  }
+
+  public async getConfigurations(accessToken: string, iModelId: string, top?: number | undefined): Promise<EC3Configuration[]> {
```
Still `iModelId` here.

Guillar1 commented on src/reporting/clients/EC3ConfigurationsClient.ts:
```diff
@@ -0,0 +1,98 @@
+        "Parameter top was outside of the valid range [1-1000]."
+      );
+    }
+    let url = `${this.basePath}/ec3/configurations?iTwinId=${iModelId}`;
```
`iTwinId=${iModelId}` 🤔

MindaugasButkus commented on src/integration-test/integration/ReportsClient.test.ts:
```diff
@@ -108,9 +113,8 @@ describe("Reports Client", () => {
-      expect(["Test1", "Test2", "Test3", "Test"]).to.include(report.displayName);
-    }
+    expect(reports.length).to.be.gt(3);
+    expect(reports.some((x) => x.deleted)).to.be.true;
```
So the test is "Get all non deleted" and you expect some reports to be deleted? Makes no sense.

RytisSapka replied:
renamed test

MindaugasButkus commented on src/test/EC3ConfigurationsClient.test.ts:
```diff
@@ -0,0 +1,229 @@
+
+  beforeEach(() => {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    fetchStub = sinon.stub(configurationsClient, "fetchJSON" as any);
```
Don't stub methods in the `beforeEach` if you are changing them in each test.

RytisSapka replied:
I am only changing them in the single test that changes the base path of the client (because a new client is created ant therefore needs to be stubbed)

MindaugasButkus commented on src/test/EC3ConfigurationsClient.test.ts:
```diff
@@ -0,0 +1,229 @@
+    };
+    fetchStub.resolves(returns);
+    await client.getConfiguration("auth", "configurationId");
+    expect(fetchStub.getCall(0).args[0]).to.match(new RegExp("^BASE"));
```
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith

I'd prefer not to use regex if it is possible.

RytisSapka replied:
comparing to substring

MindaugasButkus reviewed (changes requested):
Add new clients to `src/insights-client.ts` for them to be exported.

MindaugasButkus reviewed (changes requested):
Move EC3 clients from under `reporting` directory. EC3 clients are not for reporting, but rather for CarbonCalculation.

RytisSapka commented:
> Move EC3 clients from under `reporting` directory. EC3 clients are not for reporting, but rather for CarbonCalculation.

Reorganized. Moved shared code to common folder, renamed OneClickLCA folder to carbon-calculation and moved the new clients there.

MindaugasButkus commented on src/common/OperationsBase.ts:
```diff
@@ -3,7 +3,7 @@
 *--------------------------------------------------------------------------------------------*/
 import isomorphicFetch from "cross-fetch";
-import { DataType, ECProperty } from "./interfaces/GroupProperties";
+import { DataType, ECProperty } from "../reporting/interfaces/GroupProperties";
```
Nothing under `common` should be importing anything from `reporting` or `carbon-calculation`.

RytisSapka replied:
moved isValid function to mappingsClient since it was the only client using it.

MindaugasButkus commented on CHANGELOG.md:
```diff
@@ -0,0 +1,9 @@
+## 0.3.3
+Tue, 15 Nov 2022
+
+### Minor changes
```
```suggestion
### Patches
```

This is only a patch version bump, so I think it wouldn't be correct to call this `Minor changes`.

MindaugasButkus commented on src/test/MappingsClient.test.ts:
```diff
@@ -793,7 +792,11 @@ describe("mappingsClient", () => {
-    const privateClient = new MappingsClient() as unknown as IOperationsBase;
+    interface MappingsBase {
+      isNullOrWhitespace(input: string | null | undefined): boolean;
+      isValidECProperty (prop: ECProperty): boolean;
```
Inconsistent formatting.
